File: easy/Tree/104. Maximum Depth of Binary Tree (Stack+DFS).cpp
// https://leetcode.com/problems/maximum-depth-of-binary-tree/solutions/4254071/depth-first-search-dfs-iterative-approach-o-n-step-by-step-explanation/
// Time: O(N) - number of nodes
// Space: O(h) - the depth of the binary tree, which is h
// Stack + DFS

class Solution {
public:
    int maxDepth(TreeNode* root) {
        stack<pair<TreeNode*, int>> st;
        int ans = 0;
        st.push({root, 1});

        while (!st.empty()) {
            auto [node, depth] = st.top();
            st.pop();

            if (node) {
                ans = max(ans, depth);
                st.push({node->left, depth + 1});
                st.push({node->right, depth + 1});
            }
        }
        return ans;
    }
};

File: easy/104. Maximum Depth of Binary Tree/104. Maximum Depth of Binary Tree (Stack+DFS).py
# https://leetcode.com/problems/maximum-depth-of-binary-tree/solutions/4254071/depth-first-search-dfs-iterative-approach-o-n-step-by-step-explanation/
# Time: O(N) - number of nodes
# Space: O(h) - for the recursive stack
# Stack + DFS

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        stack = [[root , 1]]
        ans = 0

        while stack:
            node , depth = stack.pop()

            if node:
                ans = max(ans , depth)
                stack.append([node.left , depth+1])
                stack.append([node.right , depth+1])

        return ans  